### ISSUE TITLE
pnpx not working

### ISSUE THREAD

**Issue by vjpr** [pnpm/pnpm#948]
```
pnpx pg-schema-diff
Resolving: total 201, reused 176, downloaded 25, done
Adding 201 packages to node_modules
// npx: command not found: pg-schema-diff

npmx pg-schema-diff
npx: installed 202 in 12.896s
Usage: ../../.npm/_npx/5445/bin/dbdiff conn_string1 schema1 conn_string2 schema2
```

pnpm@1.23.1 

---

Looks like `pnpx` not handling a bin named something different: `dbdiff`

**Comment by vjpr** [comment]
Just ran into this again.

```
pnpx protobufjs
npx: command not found: protobufjs

npx protobufjs
usage: pbjs [options] file1.proto file2.json ...  (or pipe)  other | pbjs [options] -
```

**Comment by djalmajr** [comment]
Same with me 😞. Running in git-bash, Win10.

**Comment by tomByrer** [comment]
Same issue, Win10 Home, likely git-bash inside VSCode.
`pnpm` & `npx` both work great.

**Comment by kyeotic** [comment]
Im using wsl (Ubuntu) on Win10. Oddly pnpx finds and installs the package before telling me command not found.

```
❯nx npm-name-cli
Packages: +175
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Resolving: total 175, reused 0, downloaded 175, done

/home/tyrsius/.npm/_npx/25958/pnpm-global/2:
+ npm-name-cli 2.6.0

npx: command not found: npm-name-cli

❯npx npm-name-cli
npx: installed 175 in 10.455s
Specify one or more package names
```

**Comment by Zegnat** [comment]
Came by to report a bug with installing packages on the fly from URLs, but it looks like it may actually be a naming collision. For packages installed from URLs `pnpx` tries to run the last path segment as the bin, which obviously will not match for [things like gists](https://gist.github.com/zkat/4bc19503fe9e9309e2bfaa2c58074d32):

```bash
npx https://gist.github.com/zkat/4bc19503fe9e9309e2bfaa2c58074d32
# npx: installed 1 in 21.536s
# yay gist
pnpx https://gist.github.com/zkat/4bc19503fe9e9309e2bfaa2c58074d32
# Packages: +1
# +
# Resolving: total 1, reused 0, downloaded 1, done
#  WARN  undefined has no binaries
# 
# /Users/zegnat/.npm/_npx/29529/pnpm-global/3:
# + npx-is-cool 0.0.0
# npx: command not found: 4bc19503fe9e9309e2bfaa2c58074d32
```

**Comment by hyrious** [comment]
I just ran into this issue and tried to find out why. It turns out that pnpx is using [a fork of npx](https://github.com/pnpm/npx) and is not maintained for some time, in the mean time the npx was merged into the npm cli and there's no way to keep up with it through git.

The problem is caused by these code, where it tries to find the "matching" bin name but does not try a random one when the matching failed:

https://github.com/pnpm/npx/blob/main/index.js#L106-L111

The npx (from npm) is done by its `exec` command:

https://github.com/npm/cli/blob/latest/lib/commands/exec.js

Which guesses command name with:

https://github.com/npm/cli/blob/latest/workspaces/libnpmexec/lib/get-bin-from-manifest.js

I have no idea about what pnpm will do to fix this issue -- should we keep maintaining the pnpm/npx repo, or instead delegate pnpx to npx, or instead we use libnpmexec to make ourselves' pnpx command.

I hope @zkochan could see this issue and give some advices.
